pkg/spoke/addon/certificate_manager.go

```diff
@@ -0,0 +1,160 @@
+package addon
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	certificates "k8s.io/api/certificates/v1beta1"
+	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	certificatesinformers "k8s.io/client-go/informers/certificates/v1beta1"
+	"k8s.io/client-go/kubernetes"
+	csrclient "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+
+	"github.com/open-cluster-management/registration/pkg/clientcert"
+)
+
+// addOnCertificateManager starts contollers to manage addon client certificates
+type addOnCertificateManager struct {
+	addOnConfig
+
+	clusterName    string
+	agentName      string
+	kubeconfigData []byte
+	hubCSRInformer certificatesinformers.CertificateSigningRequestInformer
+	hubCSRClient   csrclient.CertificateSigningRequestInterface
+	kubeClient     kubernetes.Interface
+	recorder       events.Recorder
+
+	stopFunc context.CancelFunc
+}
+
+// syncAddOnConfig compares the old and new addon configs, and starts/restarts controllers if necessary
+func (m *addOnCertificateManager) syncAddOnConfig(config addOnConfig, ctx context.Context) error {
+	// no work if there is no change on addon config
+	if reflect.DeepEqual(m.addOnConfig, config) {
+		return nil
+	}
+
+	// otherwise stop the controllers started
+	if err := m.stop(); err != nil {
```

Comment:
this will be a problem since if the user add another registration, it will cause all existing secret deleted and re-register.